crates/sui-faucet/src/server.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    AppState, BatchFaucetResponse, BatchStatusFaucetResponse, FaucetConfig, FaucetError,
+    FaucetRequest, FaucetResponse, RequestMetricsLayer,
+};
+
+use axum::{
+    error_handling::HandleErrorLayer,
+    extract::Path,
+    http::StatusCode,
+    response::IntoResponse,
+    routing::{get, post},
+    BoxError, Extension, Json, Router,
+};
+use http::Method;
+use mysten_metrics::spawn_monitored_task;
+use prometheus::Registry;
+use std::{
+    borrow::Cow,
+    net::{IpAddr, SocketAddr},
+    sync::Arc,
+    time::Duration,
+};
+use std::{env, path::PathBuf};
+use sui_config::SUI_CLIENT_CONFIG;
+use sui_sdk::wallet_context::WalletContext;
+use tower::{limit::RateLimitLayer, ServiceBuilder};
+use tower_http::cors::{Any, CorsLayer};
+use tracing::{info, warn};
+use uuid::Uuid;
+
+use crate::faucet::Faucet;
+
+const CONCURRENCY_LIMIT: usize = 30;
+
+pub async fn start_faucet(
+    app_state: Arc<AppState>,
+    prometheus_registry: &Registry,
+) -> Result<(), anyhow::Error> {
+    // TODO: restrict access if needed
+    let cors = CorsLayer::new()
+        .allow_methods(vec![Method::GET, Method::POST])
+        .allow_headers(Any)
+        .allow_origin(Any);
+
+    let FaucetConfig {
+        port,
+        host_ip,
+        request_buffer_size,
+        max_request_per_second,
+        wal_retry_interval,
+        ..
+    } = app_state.config;
+
+    let max_concurrency = match env::var("MAX_CONCURRENCY") {
```

Comment:
nit: better to avoid the env lookup in library code -- rather keep it in `main.rs` and make this a parameter to `start_faucet`. Otherwise it becomes more difficult for readers/users to know what all the control surfaces are.